### PR TITLE
[MetaSchedule] Sample-Perfect-Tile

### PR DESCRIPTION
**Pull request by junrushao1994**
This PR add a schedule primitive `Sample-Perfect-Tile` that returns a random tiling plan of a given loop

### CODE REVIEW

**Comment by junrushao1994** [comment]
CC @comaniac @zxybazh

File: src/tir/schedule/primitive.h [hunk 1]
```diff
@@ -32,8 +34,8 @@ namespace tir {
  * \param max_exclusive The maximum value of the range, exclusive.
  * \return The random integer sampled in the given range.
  */
-TVM_DLL int SampleInt(support::LinearCongruentialEngine::TRandState* rand_state, int min_inclusive,
-                      int max_exclusive);
+TVM_DLL int32_t SampleInt(support::LinearCongruentialEngine::TRandState* rand_state,
+                          int32_t min_inclusive, int32_t max_exclusive);
 /*!
  * \brief Sample once category from candidates according to the probability weights.
  * \param self The schedule to update
```
Comment (zxybazh) [inline src/tir/schedule/primitive.h:41]: Could you please remove this line for me, thanks.

File: src/tir/schedule/primitive.h [hunk 2]
```diff
@@ -46,6 +48,40 @@ TVM_DLL int SampleInt(support::LinearCongruentialEngine::TRandState* rand_state,
...
+ * \param loop_sref The loop to be tiled
+ * \param n_split The number of tiles to be sampled
+ * \param max_innermost_factor The maximum tile size allowed to be sampled in the innermost loop
+ * \param decision The sampling decision
+ * \return A list of length `n`, the random perfect tile sizes sampled
+ */
+TVM_DLL std::vector<int64_t> SamplePerfectTile(
+    support::LinearCongruentialEngine::TRandState* rand_state,  //
+    const tir::StmtSRef& loop_sref, int32_t n_split, int32_t max_innermost_factor,
+    Optional<Array<Integer>>* decision);
```
Comment (zxybazh) [inline src/tir/schedule/primitive.h:84]: Is it possible that we use default argument values to avoid write 3 almost same functions?

**Review by comaniac — APPROVED** [review]
LGTM. One suggestion is maybe we could make this more general (e.g., `sample_tile(i, n=4, mode="perfect(default)|pow2|any")`). It seems not hard for your current implementation to support this feature?

**Review by zxybazh — APPROVED** [review]
LGTM. Nit added.

File: src/tir/schedule/primitive.h (hunk 2 above)
Comment (junrushao1994) [reply src/tir/schedule/primitive.h:84]: These 3 functions serve different purposes and have totally different implementations. If you check the impl, you will see how these 3 functions call each other :-)

File: src/tir/schedule/primitive.h (hunk 1 above)
Comment (junrushao1994) [reply src/tir/schedule/primitive.h:41]: We don’t need to remove this line. This API is good for internal use

**Comment by junrushao1994** [comment]
@comaniac we would probably need more deliberation in making this decision. As you may see, we have a pretty customized parameter max-innermost-factor which I’m not 100% sure proper for a general implementation. Let’s keep it as it is and add more sampling APIs if we need new functionalities (like power of 2 tile sizes)

File: src/tir/schedule/primitive.h (hunk 1 above)
Comment (zxybazh) [reply src/tir/schedule/primitive.h:41]: I mean the `self` param here is not shown in the function call so maybe we can remove it?

File: src/tir/schedule/primitive.h (hunk 2 above)
Comment (zxybazh) [reply src/tir/schedule/primitive.h:84]: Thanks for the note. I have carefully checked the implementation. The 3 function calls does differ a lot. Let's keep them as they are for conciseness.

File: src/tir/schedule/primitive.h (hunk 1 above)
Comment (junrushao1994) [reply src/tir/schedule/primitive.h:41]: oh sure, thanks for spotting this!

**Pull request merged**